explainers/signed-exchange-subresource-subtitution-explainer.md

```diff
@@ -0,0 +1,315 @@
+article.
+
+If the feed site knows that the article is depending on the JS file, the site
+can insert a link element (`<link rel="prefetch"
+href="https://cdn.publisher.example/lib.js" as="script">`) to prefetch the JS
+file while the user is browsing the feed site. But this is problematic from the
+privacy point of view. And also this doesn’t work if the UA is using
+[origin isolated HTTPCache mechanism](http://sirdarckcat.blogspot.com/2019/03/http-cache-cross-site-leaks.html).
+
+Signed Exchange has solved this
+[privacy-preserving prefetching](https://wicg.github.io/webpackage/draft-yasskin-webpackage-use-cases.html#private-prefetch)
+problem for main resources. If the publisher is providing the article in signed
+exchange format (article.html.sxg), the UA can prefetch the signed exchange from
+the feed site’s own server while the user is browsing the feed site. But there
+is no way to prefetch subresources in a privacy-preserving manner yet.
+
+Our proposal can solve this problem:
+1. The publisher provides the script file in signed exchange format (lib.js.sxg)
+along with its [header integrity value](#header-integrity-of-signed-exchange).
+1. The UA prefetches the signed exchange of the script from the feed site’s own
+server while the user is browsing the feed site.
+1. The HTTP response of the article from the publisher's server (https://publisher.example/article.html) has an allowed-alt-sxg link header to declare
+that the UA can load the script from the prefetched signed exchange.
+   ```
+   Link: <https://cdn.publisher.example/lib.js>;
+         rel="allowed-alt-sxg";
+         header-integrity="sha256-..."
+   ```
+   So, the UA can load the JS file from the prefetched signed exchange. If the
+   publisher is providing the article in signed exchange format
+   (article.html.sxg), the allowed-alt-sxg link header is in the inner response
+   headers of the signed exchange.
+
+### Reading article while offline
+If the publisher is providing the main resource and necessary subresources in
+signed exchange format, the user of the news feed site can read the article even
+while offline. This can be done using Service Worker.
+1. The Service Worker of the news feed site fetches the main resource signed
+exchange and subresources signed exchanges (maybe using
+[Background Fetch](https://wicg.github.io/background-fetch/)) while online, and
+stores them to the CacheStorage of the feed site origin.
+1. While the user is browsing the feed site, the site inserts prefetch link
+elements to prefetch the signed exchanges. The service worker responds to the
+fetch events of the prefetch with the signed exchanges in the CacheStorage.
+1. When the user clicks a link to the article’s signed exchange, the content and
+all necessary subresources are loaded from the cached signed exchange. So the
+user can read the article while offline.
+
+
+# Proposal
+
+1. Introduce SignedExchangeCache and PrefetchedSignedExchangeCache which are
```

Comment:
These seem like Chromium-internal concepts rather than things the TAG needs to think about.